### PR TITLE
Azure Blob Storage Custom Template Provider

### PR DESCRIPTION
**Pull request by kchason**
## Proposed changes
This addresses #3541 by adding a new custom template provider for downloading nuclei templates from Azure Blob Storage.

I don't see where/if the AWS S3 and GitHub providers are documented so I have not added explicit documentation for this provider, but am happy to do so if there's an appropriate place for it.

As automated testing of this function would require an authenticated service, the following manual testing was performed with a sandbox Azure environment.

**Execution with Missing Environment Variables**
nuclei - azure - missing variables

**Execution with Invalid Credentials**
nuclei - azure - invalid credentials

**Execution with Successful Download**
nuclei - azure - successful run

**Azure Blob Storage Listing**
nuclei - azure - web file listing

**Local File Listing After Execution**
Note that the `.json`, `.yaml`, and `.yml` files were downloaded, but the `.cs` file was ignored.
nuclei - azure - local directory

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works (see above)
- [x] I have added necessary documentation (if appropriate) (N/A)

### CODE REVIEW

File: v2/pkg/external/customtemplates/azure_blob.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+				} else {
+					// Increment the number of templates downloaded
+					templatesDownloaded++
+				}
+			}
+		}
+	}
+
+	// Log the number of templates downloaded
+	gologger.Info().Msgf("Downloaded %d templates from Azure Blob Storage container '%s'", templatesDownloaded, bk.containerName)
```
Comment (ehsandeep) [inline v2/pkg/external/customtemplates/azure_blob.go:75]: @kchason we can also include a downloaded path for the visibility, similar to [GitHub](https://github.com/projectdiscovery/nuclei/pull/2630/files#diff-56868292440deb9428977c5b4b034908c020c42eebdb2111a5c5d10abea9a5b7R37) cloned path.

File: v2/pkg/external/customtemplates/azure_blob.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	for pager.More() {
+		resp, err := pager.NextPage(context.TODO())
+		if err != nil {
+			gologger.Error().Msgf("Error listing templates in Azure Blob container: %v", err)
+			return
+		}
+
+		for _, blob := range resp.Segment.BlobItems {
+			// If the blob is a .yaml, .yml, or .json file, download the file to the local filesystem
+			if strings.HasSuffix(*blob.Name, ".yaml") || strings.HasSuffix(*blob.Name, ".yml") || strings.HasSuffix(*blob.Name, ".json") {
```
Comment (ehsandeep) [inline v2/pkg/external/customtemplates/azure_blob.go:61]: even `.yml` is a valid extension, nuclei only load / read `.yaml` for template loading, so we can exclude `.yml` as well.

**Review by ehsandeep — CHANGES_REQUESTED** [review]
Thank you @kchason for adding an additional provider, feel free to extend https://nuclei.projectdiscovery.io/nuclei/get-started/#custom-templates section for the docs.

File: v2/pkg/external/customtemplates/azure_blob.go (hunk 2 above)
Comment (kchason) [reply v2/pkg/external/customtemplates/azure_blob.go]: Fixed in `cd1b9d3e`

File: v2/pkg/external/customtemplates/azure_blob.go (hunk 1 above)
Comment (kchason) [reply v2/pkg/external/customtemplates/azure_blob.go]: Fixed in `b17c0a2f`

**Comment by kchason** [comment]
Added the documentation updates in https://github.com/projectdiscovery/nuclei-docs/pull/142

**Review by ehsandeep — APPROVED** [review]

**Review by ShubhamRasal — APPROVED** [review]
lgtm 
Thanks for the PR @kchason  👍🏼

**Pull request merged**